storage/pin/pin.go

```diff
@@ -0,0 +1,400 @@
+
+			// Signal end of file stream
+			close(fileWorkers)
+
+		}
+	}()
+
+	for fileRef := range fileWorkers {
+
+		// Send the file to chunk workers
+		chunkWorkers <- fileRef
+
+		actualFileSize := uint64(0)
+		rcvdFileSize := uint64(0)
+		doneChunkWorker := make(chan struct{})
+		var cwg sync.WaitGroup // Wait group to wait for chunk processing to complete
+
+	QuitChunkFor:
+		for {
+			select {
+			case <-doneChunkWorker:
+				break QuitChunkFor
+
+			case ref := <-chunkWorkers:
+				cwg.Add(1)
+
+				go func() {
+
+					defer cwg.Done()
+
+					chunkData, err := getter.Get(context.TODO(), ref)
+					if err != nil {
+						log.Error("Error getting chunk data from localstore.", "Address", hex.EncodeToString(ref))
+						close(doneChunkWorker)
+						return
+					}
+
+					datalen := len(chunkData)
+					if datalen < 9 { // Atleast 1 data byte. first 8 bytes are address
+						log.Error("Invalid chunk data from localstore.", "Address", hex.EncodeToString(ref))
+						close(doneChunkWorker)
+						return
+					}
+
+					subTreeSize := chunkData.Size()
+					if actualFileSize < subTreeSize {
+						actualFileSize = subTreeSize
+					}
+
+					if subTreeSize > chunk.DefaultSize {
+						// this is a tree chunk
+						// load the tree's branches
+						branches := (datalen - 8) / hashSize
+						for i := 0; i < branches; i++ {
+							brAddr := make([]byte, hashSize)
+							start := (i * hashSize) + 8
+							end := ((i + 1) * hashSize) + 8
+							copy(brAddr[:], chunkData[start:end])
+							chunkWorkers <- storage.Reference(brAddr)
+						}
+
+					} else {
+						// this is a data chunk
+						rcvdFileSize = rcvdFileSize + chunk.DefaultSize
+						if rcvdFileSize >= actualFileSize {
+							close(doneChunkWorker)
+						}
+					}
+
+					// process the chunk (pin / unpin / display)
+					err = executeFunc(ref)
+					if err != nil {
+						// TODO: if this happens, we should go back and revert the entire file's chunks
+						log.Error("Could not unpin chunk.", "Address", hex.EncodeToString(ref))
+					}
+				}()
+			}
+		}
+
+		// Wait for all the chunks to finish execution
+		cwg.Wait()
+
+	}
+}
+
+func (p *API) removeDecryptionKeyFromChunkHash(ref []byte) []byte {
+	// remove the decryption key from the encrypted file hash
+	isEncrypted := len(ref) > p.hashSize
+	if isEncrypted {
+		chunkAddr := make([]byte, p.hashSize)
+		copy(chunkAddr, ref[0:p.hashSize])
+		return chunkAddr
+	}
+	return ref
+}
+
+func (p *API) getPinCounterOfChunk(addr chunk.Address) (uint64, error) {
+	pinnedChunk, err := p.db.Get(context.Background(), chunk.ModeGetPin, p.removeDecryptionKeyFromChunkHash(addr))
+	if err != nil {
+		return 0, err
+	}
+	// Pin counter is passed in the Data field... decode and use it
+	return binary.BigEndian.Uint64(pinnedChunk.Data()[:8]), nil
+}
+
+func loadPinnedFilesInfo(pinnedFiles map[string]FileInfo, stateStore state.Store) error {
+	pinnedFiles = make(map[string]FileInfo)
+	err := stateStore.Get("pin-files", &pinnedFiles)
```

Comment:
As far as I see, state store is used to store only one key and one value in this case. And the value is a map of FileInfo types. We could do this only with a simple file, as well. Beside that, we are keeping all the pinned files in memory. However this may not take much memory, it is questionable, to keep two data structures one aside to another, one in memory and one for persistence. 

Why not using state store store FileInfos on hashes as store keys? This way there would be no need to keep in-memory map and deal with data races on it and when it its persisted in state store. If the performance is desired by using map, please, make sure that no data races are possible and maybe to simplify, by just storing to a simple file, instead to use state store just for one key.

Reply:
State store is used because this needs to be persisted somewhere. Without that you have to do lot of calculation every time we need information about these root hashses.

> We could do this only with a simple file, as well

Do you mean a file in file system? Will it not have the same issues?

> Beside that, we are keeping all the pinned files in memory. However this may not take much memory, it is questionable, to keep two data structures one aside to another, one in memory and one for persistence.

It's only one Data structure... The map which stores the info... and there is no other one.
When system loads up.. it read it in to memory and when there is a change it stores it in state DB. Thats it. 

> Why not using state store store FileInfos on hashes as store keys? 

The issue is that the same state store is used for multiple things.. (tags is one thing that is going to come) and we cannot guarantee they some one else will use the same hash as key and overwrite it.

The question of .. 
    What if we have tons of pinned files to store?.. 
    will it take lots of memory?.. 
    will it slow down the updates?
still remains.. 

May be we can prefix the root hash with "pin_". 
what do you think?

Reply:
> > We could do this only with a simple file, as well
> 
> Do you mean a file in file system? Will it not have the same issues?

Issues would be the same, but storing/overwriting a complete state in a possible bigger leveldb value is raising a concern for me.

> 
> > Beside that, we are keeping all the pinned files in memory. However this may not take much memory, it is questionable, to keep two data structures one aside to another, one in memory and one for persistence.
> 
> It's only one Data structure... The map which stores the info... and there is no other one.
> When system loads up.. it read it in to memory and when there is a change it stores it in state DB. Thats it.

Yes, under a single maybe large value every time something is pinned or unpinned. It slows down with the number of pinned files. Yes, writing to a file will have the same scalability issue, but it would perform much better.

> 
> > Why not using state store store FileInfos on hashes as store keys?
> 
> The issue is that the same state store is used for multiple things.. (tags is one thing that is going to come) and we cannot guarantee they some one else will use the same hash as key and overwrite it.

Maybe to construct a dedicated state store for pinning to ensure no conflicts? I believe that prefixing keys would also work acceptably reliable. I would even dare to suggest boltdb for things like this as it provides nested buckets and work quite well, but this is probably not a place for that discussion.

> 
> The question of ..
> What if we have tons of pinned files to store?..
> will it take lots of memory?..
> will it slow down the updates?
> still remains..
> 
> May be we can prefix the root hash with "pin_".
> what do you think?

Dedicated state store or prefixes, either should work fine, I think. Maybe others can provide opinions on this, too.

Reply:
> Why not using state store store FileInfos on hashes as store keys? This way there would be no need to keep in-memory map and deal with data races on it and when it its persisted in state store.

exactly so. Lets simplify please, use state store with addresses as keys and no in-memory representation needed. 
This store is accessed very rarely and surely not time critical

Reply:
Using state store now with keys prefixed with "pin_". No in memory structure is used.